contributors/design-proposals/storage/csi-migration.md

```diff
@@ -661,6 +721,111 @@ with minimal disruption to users would be:
+as detailed below.
+
+### Enhancements in Probing/Registration of in-tree plugins
+
+Functions appendLegacyProviderVolumes (in the Kubernetes Controller Manager and Kubelet)
+and appendAttachableLegacyProviderVolumes (in the Kubernetes Controller Manager)
+will be enhanced to [1] check plugin specific migration feature flags as well as
+[2] the overall CSIMigrationInTreeOff feature flag to determine whether ProbeVolumePlugins
+function of a legacy in-tree plugin will get invoked.
+
+Once CSIMigrationInTreeOff feature flag and the plugin specific migration
+flags get enabled by default, the build tag `nolegacyproviders` can be enabled on
+that no in-tree plugins get probed.
+
+### Detection of migration status for a plugin
+
+Code paths that need to check migration status of a plugin, for example,
+in provisionClaimOperationExternal, findDeletablePlugin, etc. will need to
+depend on an instance of the CSIMigratedPluginManager that provides the following
+utilities:
+```
+func (pm CSIMigratedPluginManager) IsCSIMigrationEnabledForPluginByName(pluginName string) bool
+func (pm CSIMigratedPluginManager) IsPluginMigratableToCSIBySpec(spec *Spec) (bool, error)
+```
+The CSIMigratedPluginManager will be introduced in pkg/volume/csi_migration.go
+
+Note that a per-plugin member function of the form IsMigratedToCSI cannot be used
+since [1] a plugin object for in-tree plugins will typically be nil and [2] the
+code implementing IsMigratedToCSI will be removed as part of disabling plugin code.
+
+### Handling of errors returned by FindPluginBySpec/FindPluginByName for legacy plugins
+
+Once an in-tree plugin is no longer probed (through ProbeVolumePlugins), all the VolumePluginMgr
+functions of the form Find*PluginBySpec/Find*PluginByName will return an error.
+For example, invocation of:
+
+1. FindProvisionablePluginByName in the pv controller will return error for a migrated
+plugin that is no longer probed in ProbeControllerVolumePlugins.
+2. FindExpandablePluginBySpec in the expand controller will return error for a migrated
+plugin that is no longer probed in ProbeExpandableVolumePlugins.
+3. FindAttachablePluginBySpec in the attach/detach controller will return error for
+a migrated plugin that is no longer probed in ProbeAttachableVolumePlugins.
+
+Code invoking the above functions will need to check for migration status of a plugin
+based on Volume spec or Plugin name rather than throwing an error.
+
+### Enhancements in Controllers handling Persistent Volumes
+
+#### AttachDetach Controller
+The AttachDetach Controller will translate volume specs for an in-tree plugin to
+a migrated CSI plugin at the points where Desired State of World cache gets
+populated through CreateVolumeSpec when the following conditions are true:
+[1] CSIMigration or CSIMigrationInTreeOff feature flag is enabled.
+[2] A plugin-specific migration feature flag is enabled. 
+[3] A node has a migrated CSI plugin installed that corresponds to an
```

Comment:
it might be worth having a small section that explains when the new flag "InTreeMigrationOff=true" that the Kubelet *MUST* have migration on (since there will be no more fall-back possibility).

The implementation consequence is that `nodeUsingCSIPlugin` can simply `return true` if `CSIMigrationInTreeOff=true`